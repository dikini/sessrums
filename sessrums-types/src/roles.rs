//! Role definitions for session type protocols.
//! 
//! Roles are implemented as zero-sized types (ZSTs) to provide type-level
//! identification of protocol participants without runtime overhead.

/// Trait marking types that can act as roles in a protocol.
/// 
/// This trait is sealed and cannot be implemented outside this crate
/// to ensure role type safety.
pub trait Role: private::Sealed {}

/// Client role in a client-server protocol.
#[derive(Debug, Clone, Copy)]
pub struct Client;

/// Server role in a client-server protocol.
#[derive(Debug, Clone, Copy)]
pub struct Server;

// Implement Role for our basic roles
impl Role for Client {}
impl Role for Server {}

// Private sealed trait to prevent external implementations
mod private {
    pub trait Sealed {}
    impl Sealed for super::Client {}
    impl Sealed for super::Server {}
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roles_are_copy() {
        let client = Client;
        let _client_copy = client;
        // Original client still usable
        let _also_client = client;
        
        let server = Server;
        let _server_copy = server;
        // Original server still usable
        let _also_server = server;
    }
}