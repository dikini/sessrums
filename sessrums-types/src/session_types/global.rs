//! Global protocol representation for multiparty session types.
//!
//! This module provides types for representing the global view of a multiparty
//! session type protocol, where interactions between all participants are
//! described from a centralized perspective.

use std::marker::PhantomData;
use std::collections::HashSet;
use super::common::{RoleIdentifier, Label, RecursionLabel};

/// Represents a global protocol interaction in a multiparty session type system.
///
/// `GlobalInteraction` provides a way to describe the global view of a protocol,
/// where all interactions between participants are specified from a centralized
/// perspective. This is in contrast to the local (or endpoint) view, where each
/// participant has their own perspective of the protocol.
///
/// The global view is essential for:
/// 1. Protocol design and specification
/// 2. Verification of protocol properties (e.g., deadlock freedom)
/// 3. Projection to local types for each participant
/// 4. Runtime monitoring and enforcement of protocol compliance
///
/// # Type Parameters
///
/// * `M` - A generic type parameter representing the message type
///
/// # Variants
///
/// * `Message` - Represents a message exchange from one role to another
/// * `Choice` - Represents a choice point where one role selects a branch
/// * `Rec` - Represents a recursion point in the protocol
/// * `Var` - Represents a reference to a recursion point
/// * `End` - Represents the termination of the protocol
///
/// # Examples
///
/// ```
/// use std::marker::PhantomData;
/// use sessrums_types::session_types::global::GlobalInteraction;
/// use sessrums_types::session_types::common::RoleIdentifier;
///
/// // Define a simple protocol: client sends a request to server, server responds
/// let protocol: GlobalInteraction<String> = GlobalInteraction::Message {
///     from: RoleIdentifier::new("client"),
///     to: RoleIdentifier::new("server"),
///     msg: PhantomData::<String>,
///     cont: Box::new(GlobalInteraction::Message {
///         from: RoleIdentifier::new("server"),
///         to: RoleIdentifier::new("client"),
///         msg: PhantomData::<String>,
///         cont: Box::new(GlobalInteraction::End),
///     }),
/// };
/// ```
#[derive(Debug, Clone)]
pub enum GlobalInteraction<M: Clone> {
    /// Represents a message exchange from one role to another.
    ///
    /// # Fields
    ///
    /// * `from` - The role identifier of the sender
    /// * `to` - The role identifier of the receiver
    /// * `msg` - A phantom type representing the message type
    /// * `cont` - The continuation of the protocol after this message exchange
    Message {
        /// The role identifier of the sender
        from: RoleIdentifier,
        
        /// The role identifier of the receiver
        to: RoleIdentifier,
        
        /// A phantom type representing the message type
        msg: PhantomData<M>,
        
        /// The continuation of the protocol after this message exchange
        cont: Box<GlobalInteraction<M>>,
    },
    
    /// A choice point in the protocol, where the deciding role selects one of the branches.
    ///
    /// # Fields
    ///
    /// * `decider` - The role that makes the decision
    /// * `branches` - The available branches, each with a label and continuation
    ///
    /// # Examples
    ///
    /// ```
    /// use std::marker::PhantomData;
    /// use sessrums_types::session_types::global::GlobalInteraction;
    /// use sessrums_types::session_types::common::{RoleIdentifier, Label};
    ///
    /// // Define a protocol with a choice: client chooses between "login" and "register"
    /// let protocol = GlobalInteraction::choice(
    ///     "client",
    ///     vec![
    ///         ("login".into(), GlobalInteraction::message(
    ///             "client",
    ///             "server",
    ///             GlobalInteraction::end(),
    ///         )),
    ///         ("register".into(), GlobalInteraction::message(
    ///             "client",
    ///             "server",
    ///             GlobalInteraction::end(),
    ///         )),
    ///     ],
    /// );
    /// ```
    Choice {
        /// The role that makes the decision
        decider: RoleIdentifier,
        
        /// The available branches, each with a label and continuation
        branches: Vec<(Label, Box<GlobalInteraction<M>>)>,
    },
    
    /// Recursion point in the protocol.
    ///
    /// # Fields
    ///
    /// * `label` - The label identifying this recursion point
    /// * `body` - The body of the recursive protocol
    ///
    /// # Examples
    ///
    /// ```
    /// use std::marker::PhantomData;
    /// use sessrums_types::session_types::global::GlobalInteraction;
    /// use sessrums_types::session_types::common::{RoleIdentifier, RecursionLabel};
    ///
    /// // Define a recursive ping-pong protocol
    /// let protocol = GlobalInteraction::rec(
    ///     "loop",
    ///     GlobalInteraction::message(
    ///         "client",
    ///         "server",
    ///         GlobalInteraction::message(
    ///             "server",
    ///             "client",
    ///             GlobalInteraction::var("loop"),
    ///         ),
    ///     ),
    /// );
    /// ```
    Rec {
        /// The label identifying this recursion point
        label: RecursionLabel,
        
        /// The body of the recursive protocol
        body: Box<GlobalInteraction<M>>,
    },
    
    /// Reference to a recursion point.
    ///
    /// # Fields
    ///
    /// * `label` - The label of the referenced recursion point
    Var {
        /// The label of the referenced recursion point
        label: RecursionLabel,
    },
    
    /// Represents the termination of the protocol.
    ///
    /// When a protocol reaches the `End` state, no further interactions are expected.
    End,
}

impl<M: Clone> GlobalInteraction<M> {
    /// Creates a new message interaction in the global protocol.
    ///
    /// # Parameters
    ///
    /// * `from` - The role identifier of the sender
    /// * `to` - The role identifier of the receiver
    /// * `cont` - The continuation of the protocol after this message exchange
    ///
    /// # Returns
    ///
    /// A new `GlobalInteraction::Message` variant
    pub fn message(
        from: impl Into<RoleIdentifier>,
        to: impl Into<RoleIdentifier>,
        cont: GlobalInteraction<M>,
    ) -> Self {
        GlobalInteraction::Message {
            from: from.into(),
            to: to.into(),
            msg: PhantomData,
            cont: Box::new(cont),
        }
    }

    /// Creates a new end interaction, representing protocol termination.
    ///
    /// # Returns
    ///
    /// A new `GlobalInteraction::End` variant
    pub fn end() -> Self {
        GlobalInteraction::End
    }
    
    /// Creates a new choice interaction in the global protocol.
    ///
    /// # Parameters
    ///
    /// * `decider` - The role identifier of the participant making the choice
    /// * `branches` - A vector of labeled branches, each with a continuation
    ///
    /// # Returns
    ///
    /// A new `GlobalInteraction::Choice` variant
    pub fn choice(
        decider: impl Into<RoleIdentifier>,
        branches: Vec<(Label, GlobalInteraction<M>)>,
    ) -> Self {
        GlobalInteraction::Choice {
            decider: decider.into(),
            branches: branches
                .into_iter()
                .map(|(label, cont)| (label, Box::new(cont)))
                .collect(),
        }
    }
    
    /// Creates a new recursion point in the global protocol.
    ///
    /// # Parameters
    ///
    /// * `label` - The label identifying this recursion point
    /// * `body` - The body of the recursive protocol
    ///
    /// # Returns
    ///
    /// A new `GlobalInteraction::Rec` variant
    pub fn rec(
        label: impl Into<RecursionLabel>,
        body: GlobalInteraction<M>,
    ) -> Self {
        GlobalInteraction::Rec {
            label: label.into(),
            body: Box::new(body),
        }
    }
    
    /// Creates a new reference to a recursion point.
    ///
    /// # Parameters
    ///
    /// * `label` - The label of the referenced recursion point
    ///
    /// # Returns
    ///
    /// A new `GlobalInteraction::Var` variant
    pub fn var(label: impl Into<RecursionLabel>) -> Self {
        GlobalInteraction::Var {
            label: label.into(),
        }
    }
    
    /// Check if this protocol is well-formed with respect to recursion.
    ///
    /// This method verifies that:
    /// 1. All `Var` references point to defined `Rec` labels
    /// 2. Recursion is properly nested
    ///
    /// # Returns
    ///
    /// `Ok(())` if the protocol is well-formed, or an error message if not
    pub fn check_recursion_well_formedness(&self) -> Result<(), String> {
        let mut defined_labels = HashSet::new();
        self.check_recursion_well_formedness_inner(&mut defined_labels)
    }
    
    fn check_recursion_well_formedness_inner(
        &self, 
        defined_labels: &mut HashSet<RecursionLabel>
    ) -> Result<(), String> {
        match self {
            GlobalInteraction::Message { cont, .. } => {
                cont.check_recursion_well_formedness_inner(defined_labels)
            },
            
            GlobalInteraction::Choice { branches, .. } => {
                for (_, branch) in branches {
                    branch.check_recursion_well_formedness_inner(defined_labels)?;
                }
                Ok(())
            },
            
            GlobalInteraction::Rec { label, body } => {
                if defined_labels.contains(label) {
                    return Err(format!("Duplicate recursion label: {:?}", label));
                }
                defined_labels.insert(label.clone());
                let result = body.check_recursion_well_formedness_inner(defined_labels);
                defined_labels.remove(label);
                result
            },
            
            GlobalInteraction::Var { label } => {
                if !defined_labels.contains(label) {
                    return Err(format!("Reference to undefined recursion label: {:?}", label));
                }
                Ok(())
            },
            
            GlobalInteraction::End => Ok(()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_simple_protocol() {
        // Define a simple protocol: client sends a request to server, server responds
        let protocol: GlobalInteraction<String> = GlobalInteraction::message(
            "client",
            "server",
            GlobalInteraction::message(
                "server",
                "client",
                GlobalInteraction::end(),
            ),
        );

        // Verify the structure of the protocol
        if let GlobalInteraction::Message { from, to, cont, .. } = protocol {
            assert_eq!(from.name(), "client");
            assert_eq!(to.name(), "server");
            
            if let GlobalInteraction::Message { from, to, cont, .. } = *cont {
                assert_eq!(from.name(), "server");
                assert_eq!(to.name(), "client");
                
                assert!(matches!(*cont, GlobalInteraction::End));
            } else {
                panic!("Expected Message, got End");
            }
        } else {
            panic!("Expected Message, got End");
        }
    }

    #[test]
    fn test_create_three_party_protocol() {
        // Define a three-party protocol: client sends to server, server forwards to database, database responds to server, server responds to client
        let protocol: GlobalInteraction<String> = GlobalInteraction::message(
            "client",
            "server",
            GlobalInteraction::message(
                "server",
                "database",
                GlobalInteraction::message(
                    "database",
                    "server",
                    GlobalInteraction::message(
                        "server",
                        "client",
                        GlobalInteraction::end(),
                    ),
                ),
            ),
        );

        // Verify the first interaction
        if let GlobalInteraction::Message { from, to, cont, .. } = protocol {
            assert_eq!(from.name(), "client");
            assert_eq!(to.name(), "server");
            
            // Verify the second interaction
            if let GlobalInteraction::Message { from, to, cont, .. } = *cont {
                assert_eq!(from.name(), "server");
                assert_eq!(to.name(), "database");
                
                // Verify the third interaction
                if let GlobalInteraction::Message { from, to, cont, .. } = *cont {
                    assert_eq!(from.name(), "database");
                    assert_eq!(to.name(), "server");
                    
                    // Verify the fourth interaction
                    if let GlobalInteraction::Message { from, to, cont, .. } = *cont {
                        assert_eq!(from.name(), "server");
                        assert_eq!(to.name(), "client");
                        
                        // Verify the end
                        assert!(matches!(*cont, GlobalInteraction::End));
                    } else {
                        panic!("Expected fourth Message, got End");
                    }
                } else {
                    panic!("Expected third Message, got End");
                }
            } else {
                panic!("Expected second Message, got End");
            }
        } else {
            panic!("Expected first Message, got End");
        }
    }
    
    #[test]
    fn test_create_recursive_protocol() {
        // Define a recursive ping-pong protocol
        let protocol: GlobalInteraction<String> = GlobalInteraction::rec(
            "loop",
            GlobalInteraction::message(
                "client",
                "server",
                GlobalInteraction::message(
                    "server",
                    "client",
                    GlobalInteraction::var("loop"),
                ),
            ),
        );
        
        // Verify the structure of the protocol
        if let GlobalInteraction::Rec { label, body } = protocol {
            assert_eq!(label.name(), "loop");
            
            if let GlobalInteraction::Message { from, to, cont, .. } = *body {
                assert_eq!(from.name(), "client");
                assert_eq!(to.name(), "server");
                
                if let GlobalInteraction::Message { from, to, cont, .. } = *cont {
                    assert_eq!(from.name(), "server");
                    assert_eq!(to.name(), "client");
                    
                    if let GlobalInteraction::Var { label } = *cont {
                        assert_eq!(label.name(), "loop");
                    } else {
                        panic!("Expected Var, got something else");
                    }
                } else {
                    panic!("Expected Message, got something else");
                }
            } else {
                panic!("Expected Message, got something else");
            }
        } else {
            panic!("Expected Rec, got something else");
        }
        
        // Verify well-formedness
        assert!(protocol.check_recursion_well_formedness().is_ok());
    }
    
    #[test]
    fn test_invalid_recursive_protocol() {
        // Define a protocol with an undefined recursion variable
        let protocol: GlobalInteraction<String> = GlobalInteraction::message(
            "client",
            "server",
            GlobalInteraction::var("undefined"),
        );
        
        // Verify that well-formedness check fails
        assert!(protocol.check_recursion_well_formedness().is_err());
        
        // Define a protocol with duplicate recursion labels
        let protocol: GlobalInteraction<String> = GlobalInteraction::rec(
            "loop",
            GlobalInteraction::rec(
                "loop",
                GlobalInteraction::end(),
            ),
        );
        
        // Verify that well-formedness check fails
        assert!(protocol.check_recursion_well_formedness().is_err());
    }
}