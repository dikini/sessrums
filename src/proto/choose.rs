//! Choose protocol type for session types.
//!
//! This module defines the `Choose<L, R>` protocol type, which represents a protocol
//! that chooses between two continuations, `L` and `R`.

use std::marker::PhantomData;
use super::Protocol;
use super::offer::Offer;

/// A protocol that chooses between two continuations, `L` and `R`.
///
/// The `Choose<L, R>` type represents a protocol that chooses between
/// two continuations, `L` and `R`. The choice is made by this party.
///
/// # Duality
///
/// The dual of `Choose<L, R>` is `Offer<L::Dual, R::Dual>`. This reflects the fundamental
/// session type principle that if one party chooses, the other party must offer the options.
///
/// This duality relationship ensures protocol compatibility between communicating parties:
/// - When one process chooses between protocols L and R, the other process must
///   offer the dual protocols L::Dual and R::Dual.
/// - The duality relationship is symmetric: `Choose<L, R>::Dual::Dual == Choose<L, R>`.
///
/// ## Duality Transformation
///
/// The duality transformation for Choose follows these rules:
/// 1. Replace Choose with Offer
/// 2. Apply the duality transformation to each branch (L becomes L::Dual, R becomes R::Dual)
///
/// For example:
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Offer, Choose};
///
/// // Original protocol
/// type MyChoose = Choose<Send<i32, End>, Recv<String, End>>;
///
/// // Its dual
/// type MyChooseDual = Offer<Recv<i32, End>, Send<String, End>>;
/// ```
///
/// ## Nested Duality
///
/// The duality relationship extends to nested Choose types:
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Offer, Choose};
///
/// // A protocol with nested Choose
/// type NestedChoose = Choose<
///     Choose<Send<i32, End>, Recv<bool, End>>,
///     Send<String, End>
/// >;
///
/// // Its dual has nested Offer
/// type NestedChooseDual = Offer<
///     Offer<Recv<i32, End>, Send<bool, End>>,
///     Recv<String, End>
/// >;
/// ```
///
/// # Type Parameters
///
/// * `L` - The first continuation protocol type. Must implement the `Protocol` trait.
/// * `R` - The second continuation protocol type. Must implement the `Protocol` trait.
///
/// # Examples
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Choose};
///
/// // A protocol that chooses between:
/// // 1. Sending an i32 and then ending, or
/// // 2. Receiving a String and then ending
/// type MyProtocol = Choose<Send<i32, End>, Recv<String, End>>;
///
/// // The dual protocol would offer:
/// // 1. Receiving an i32 and then ending, or
/// // 2. Sending a String and then ending
/// // (This would be: Offer<Recv<i32, End>, Send<String, End>>)
/// ```
///
/// # Protocol Composition
///
/// `Choose<L, R>` can be composed with other protocol types to create more complex
/// communication patterns:
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Choose};
///
/// // A protocol that first sends a boolean, then chooses between
/// // sending an i32 or receiving a String
/// type ComplexProtocol = Send<bool, Choose<Send<i32, End>, Recv<String, End>>>;
/// ```
pub struct Choose<L, R> {
    _marker: PhantomData<(L, R)>,
}

impl<L: Protocol, R: Protocol> Protocol for Choose<L, R> {
    /// The dual of `Choose<L, R>` is `Offer<L::Dual, R::Dual>`.
    ///
    /// This reflects the fact that if one party chooses, the other party
    /// must offer the options.
    type Dual = Offer<L::Dual, R::Dual>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proto::{Send, Recv, End};

    #[test]
    fn test_choose_protocol_implementation() {
        // This test verifies that Choose<L, R> implements Protocol
        
        // We can't directly test the equality of types in Rust,
        // but we can verify that Choose<L, R> implements Protocol
        fn assert_implements_protocol<T: Protocol>() {}
        
        // This will compile only if Choose<Send<i32, End>, Recv<String, End>> implements Protocol
        assert_implements_protocol::<Choose<Send<i32, End>, Recv<String, End>>>();
    }

    #[test]
    fn test_choose_duality() {
        // This test verifies the duality relationship for Choose<L, R>
        
        // Define some simple protocol types for testing
        type ChooseProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        type ExpectedDual = Offer<Recv<i32, End>, Send<String, End>>;
        
        // Check that ChooseProtocol::Dual is ExpectedDual
        fn check_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if ChooseProtocol::Dual is ExpectedDual
        check_dual::<ChooseProtocol, ExpectedDual>();
    }

    #[test]
    fn test_choose_duality_symmetry() {
        // This test verifies the symmetry of the duality relationship for Choose<L, R>
        // Choose<L, R>::Dual is Offer<L::Dual, R::Dual>
        // Offer<L::Dual, R::Dual>::Dual should be Choose<L, R>
        
        // Define some simple protocol types for testing
        type ChooseProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        type DualProtocol = Offer<Recv<i32, End>, Send<String, End>>;
        
        // Check that the duality relationship is symmetric
        fn check_duality_symmetry<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the duality relationship is correctly implemented
        check_duality_symmetry::<ChooseProtocol, DualProtocol>();
    }

    #[test]
    fn test_choose_with_complex_composition() {
        // Test Choose in a more complex protocol composition
        
        // A protocol that sends a boolean, then chooses between
        // sending an i32 or receiving a String
        type ComplexProtocol = Send<bool, Choose<Send<i32, End>, Recv<String, End>>>;
        
        // The dual should be Recv<bool, Offer<Recv<i32, End>, Send<String, End>>>
        type ExpectedDual = Recv<bool, Offer<Recv<i32, End>, Send<String, End>>>;
        
        // Check that the dual relationships are correctly established
        fn check_complex_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if ComplexProtocol's dual is correctly derived
        check_complex_dual::<ComplexProtocol, ExpectedDual>();
    }

    #[test]
    fn test_nested_choose_duality() {
        // Test duality with nested Choose types
        
        // A protocol that chooses between:
        // 1. Choosing between Send<i32, End> and Recv<bool, End>
        // 2. Sending a String and then ending
        type NestedChooseProtocol = Choose<
            Choose<Send<i32, End>, Recv<bool, End>>,
            Send<String, End>
        >;
        
        // The dual should be Offer between:
        // 1. Offering between Recv<i32, End> and Send<bool, End>
        // 2. Receiving a String and then ending
        type ExpectedDual = Offer<
            Offer<Recv<i32, End>, Send<bool, End>>,
            Recv<String, End>
        >;
        
        // Check that the dual relationships are correctly established
        fn check_nested_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if NestedChooseProtocol's dual is correctly derived
        check_nested_dual::<NestedChooseProtocol, ExpectedDual>();
    }

    #[test]
    fn test_multiple_level_duality() {
        // Test multiple levels of duality (dual of dual of dual)
        
        // Start with a simple protocol
        type OriginalProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        
        // Get its dual
        type FirstDual = <OriginalProtocol as Protocol>::Dual;
        
        // Get the dual of the dual
        type SecondDual = <FirstDual as Protocol>::Dual;
        
        // Get the dual of the dual of the dual
        type ThirdDual = <SecondDual as Protocol>::Dual;
        
        // The dual of the dual should be the original protocol
        fn check_dual_of_dual<P, D, DD>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = DD>,
            DD: Protocol,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the dual of the dual is the original protocol
        check_dual_of_dual::<OriginalProtocol, FirstDual, SecondDual>();
        
        // The dual of the dual of the dual should be the dual of the original protocol
        fn check_dual_of_dual_of_dual<P, D, DD, DDD>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = DD>,
            DD: Protocol<Dual = DDD>,
            DDD: Protocol,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the dual of the dual of the dual is the dual of the original protocol
        check_dual_of_dual_of_dual::<OriginalProtocol, FirstDual, SecondDual, ThirdDual>();
    }

    #[test]
    fn test_mixed_protocol_composition() {
        // Test a mixed composition of Choose and other protocol types
        
        // A protocol with Choose nested within other protocol types
        type MixedProtocol = Send<bool,
            Choose<
                Recv<i32, End>,
                Send<String,
                    Choose<
                        Recv<char, End>,
                        Recv<u8, End>
                    >
                >
            >
        >;
        
        // The dual should have Send/Recv swapped and Choose/Offer swapped
        type ExpectedDual = Recv<bool,
            Offer<
                Send<i32, End>,
                Recv<String,
                    Offer<
                        Send<char, End>,
                        Send<u8, End>
                    >
                >
            >
        >;
        
        // Check that the dual relationships are correctly established
        fn check_mixed_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if MixedProtocol's dual is correctly derived
        check_mixed_dual::<MixedProtocol, ExpectedDual>();
    }
}