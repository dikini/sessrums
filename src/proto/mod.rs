//! Protocol type definitions for session types.
//!
//! This module contains the core protocol type definitions used to express
//! communication protocols at the type level.

pub mod roles;
pub use roles::{Role, RoleA, RoleB};

mod proto;
pub use proto::Protocol;

mod send;
pub use send::Send;

mod recv;
pub use recv::Recv;

mod end;
pub use end::End;

// Phase 2 implementations
mod offer;
pub use offer::Offer;

mod choose;
// Choose is now fully implemented in Task 2.4
pub use choose::Choose;

// These will be uncommented in Phase 5
mod rec;
pub use rec::Rec;

mod var;
pub use var::Var;

pub mod global;
pub use global::{
    GlobalProtocol, GSend, GRecv, GChoice, GOffer, GRec, GVar, GEnd, GSeq, GPar,
    GlobalProtocolBuilder, validate_global_protocol
};

pub mod projection;
pub use projection::{Project, project};

// Phase 5 implementations
pub mod compat;
pub use compat::{ProtocolCompat, BinaryWrapper, MPSTWrapper};

// Re-export the global_protocol macro from the sessrums-macro crate
pub use sessrums_macro::global_protocol;