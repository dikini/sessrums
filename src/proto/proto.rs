//! Core Protocol trait for session types.
//!
//! This module defines the fundamental `Protocol` trait that serves as the
//! foundation for the session type system. Session types allow for compile-time
//! verification of communication protocol adherence between communicating parties.

/// The `Protocol` trait represents a communication protocol type.
///
/// In session type theory, protocols describe the communication behavior between
/// two parties. Each protocol has a dual counterpart that represents the behavior
/// of the other party in the communication.
///
/// # Duality
///
/// Duality is a fundamental concept in session types. For every protocol `P`,
/// there exists a dual protocol `P::Dual` that represents the complementary
/// behavior. For example:
///
/// - If `P` sends a message of type `T`, then `P::Dual` receives a message of type `T`.
/// - If `P` offers a choice between protocols, then `P::Dual` makes a choice between the dual protocols.
///
/// This duality ensures that when two parties follow dual protocols, their
/// communication is guaranteed to be compatible and free from communication errors
/// like deadlocks or protocol violations.
///
/// # Examples
///
/// ```
/// use sez::proto::{Protocol, Send, Recv, End};
///
/// // A protocol that sends an i32, then receives a String, then ends
/// type MyProtocol = Send<i32, Recv<String, End>>;
///
/// // The dual protocol receives an i32, then sends a String, then ends
/// // (This will be implemented in future tasks)
/// // type MyDualProtocol = <MyProtocol as Protocol>::Dual;
/// // This would be equivalent to: Recv<i32, Send<String, End>>
/// ```
///
/// # Type Safety
///
/// The `Protocol` trait and its implementations leverage Rust's type system to
/// ensure protocol adherence at compile time. This means that protocol violations
/// are caught as type errors during compilation, preventing runtime communication
/// errors.
pub trait Protocol: Sized {
    /// The dual protocol type.
    ///
    /// For any protocol `P`, `P::Dual` represents the complementary protocol
    /// that can communicate with `P` without errors.
    ///
    /// The duality relationship is symmetric, meaning that for any protocol `P`:
    /// `P::Dual::Dual == P`
    type Dual: Protocol;
}

#[cfg(test)]
mod tests {
    use super::*;

    // Define some simple protocol types for testing
    struct TestProtocol;
    struct DualTestProtocol;

    // Implement Protocol for TestProtocol
    impl Protocol for TestProtocol {
        type Dual = DualTestProtocol;
    }

    // Implement Protocol for DualTestProtocol
    impl Protocol for DualTestProtocol {
        type Dual = TestProtocol;
    }

    #[test]
    fn test_protocol_duality_symmetry() {
        // This test verifies the symmetry of the duality relationship
        // by checking that the dual of the dual is the original type.
        
        // We can't directly compare types in Rust, but we can use type inference
        // to verify that the types match.
        
        // TestProtocol::Dual is DualTestProtocol
        // DualTestProtocol::Dual is TestProtocol
        // So TestProtocol::Dual::Dual is TestProtocol, which confirms the symmetry property.
        
        // This is a compile-time check - if these types match up correctly,
        // the code will compile successfully.
        fn check_dual_symmetry<P, Q>()
        where
            P: Protocol<Dual = Q>,
            Q: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // Verify that TestProtocol and DualTestProtocol are duals of each other
        check_dual_symmetry::<TestProtocol, DualTestProtocol>();
    }
}