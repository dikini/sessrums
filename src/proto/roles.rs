//! Defines the `Role` trait and concrete role types for MPST protocols.

/// A trait representing a role in an MPST protocol.
///
/// Each participant in a protocol is assigned a unique role. Roles are used
/// to distinguish between different participants and to define the structure
/// of the session types from each participant's perspective.
pub trait Role: Send + 'static + Default {
    /// Returns a string representation of the role.
    fn name(&self) -> &'static str;
}

/// A concrete role type, RoleA.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct RoleA;

impl Default for RoleA {
    fn default() -> Self {
        RoleA
    }
}

impl Role for RoleA {
    fn name(&self) -> &'static str {
        "RoleA"
    }
}

/// A concrete role type, RoleB.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct RoleB;

impl Default for RoleB {
    fn default() -> Self {
        RoleB
    }
}

impl Role for RoleB {
    fn name(&self) -> &'static str {
        "RoleB"
    }
}

// Add more concrete role types as needed for specific protocols.

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_role_a_name() {
        let role_a = RoleA;
        assert_eq!(role_a.name(), "RoleA");
    }

    #[test]
    fn test_role_b_name() {
        let role_b = RoleB;
        assert_eq!(role_b.name(), "RoleB");
    }

    #[test]
    fn test_role_a_debug() {
        let role_a = RoleA;
        assert_eq!(format!("{:?}", role_a), "RoleA");
    }

    #[test]
    fn test_role_b_debug() {
        let role_b = RoleB;
        assert_eq!(format!("{:?}", role_b), "RoleB");
    }

    #[test]
    fn test_role_a_clone() {
        let role_a = RoleA;
        let cloned_role_a = role_a.clone();
        assert_eq!(role_a, cloned_role_a);
    }

    #[test]
    fn test_role_b_clone() {
        let role_b = RoleB;
        let cloned_role_b = role_b.clone();
        assert_eq!(role_b, cloned_role_b);
    }

    #[test]
    fn test_role_a_partial_eq() {
        let role_a1 = RoleA;
        let role_a2 = RoleA;
        assert_eq!(role_a1, role_a2);
    }

    #[test]
    fn test_role_b_partial_eq() {
        let role_b1 = RoleB;
        let role_b2 = RoleB;
        assert_eq!(role_b1, role_b2);
    }

    #[test]
    fn test_role_a_b_not_equal() {
        let role_a = RoleA;
        let role_b = RoleB;
        // We can't directly compare different types with assert_ne!
        // Instead, let's compare their names
        assert_ne!(role_a.name(), role_b.name());
    }
}