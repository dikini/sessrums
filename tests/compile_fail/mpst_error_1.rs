//! Test that a global protocol with the same role as sender and receiver fails to compile.

use sessrums::proto::global::{GSend, GEnd};
use sessrums::proto::roles::{Role, RoleA};

// Define a role
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
struct Client;

impl Role for Client {
    fn name(&self) -> &'static str {
        "Client"
    }
}

fn main() {
    // This should fail to compile because the sender and receiver are the same role
    let protocol = GSend::<String, Client, Client, GEnd>::new();
    protocol.validate().unwrap(); // This should fail at compile time
}