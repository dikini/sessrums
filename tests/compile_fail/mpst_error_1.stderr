error: Sender and receiver must be different roles in GSend
 --> $DIR/mpst_error_1.rs:17:5
  |
17 |     let protocol = GSend::<String, Client, Client, GEnd>::new();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the method `validate` (in Nightly builds, run with -Z macro-backtrace for more info)